File: elixir/worg-agent/lib/worg_agent/scheduler.ex
defmodule WorgAgent.Scheduler do
  @moduledoc """
  Topological wave-planner for task DAGs. Given a list of tasks with
  declared dependencies, produces a sequence of waves where every
  task in a wave can run concurrently (all its dependencies are
  satisfied by earlier waves). Pure data transformation — no
  concurrency primitives, no side effects, no Oban.

  Pairs with the worg-agent Loop, which today executes tasks
  serially. The Scheduler's role is the *plan*; the integration that
  *dispatches* a wave via Task.async_stream / Oban / equivalent is
  the next ticket (see follow-up filed under wb-lrul). This module
  ships the foundation: deterministic wave assignment, cycle
  detection, and a `next_ready/2` predicate the Loop can use once
  integration lands.

  wb-lrul (primitive only — Loop integration deferred).

  ## Use

      iex> tasks = [
      ...>   {"research", []},
      ...>   {"script", ["research"]},
      ...>   {"velocity", ["research"]},
      ...>   {"storyboard", ["script", "velocity"]},
      ...>   {"shots", ["storyboard"]}
      ...> ]
      iex> WorgAgent.Scheduler.topological_waves(tasks)
      {:ok, [["research"], ["script", "velocity"], ["storyboard"], ["shots"]]}

  Wave 2 (`script` + `velocity`) demonstrates the speedup: both can
  fan out concurrently once research lands.

  ## Cycle detection

  If the task list contains a cycle (`a → b → a`), returns
  `{:error, {:cycle, [task_ids_in_cycle]}}`. No partial result.

  ## Unknown dependency detection

  If a task lists a `depends_on` id that isn't in the task list,
  returns `{:error, {:unknown_dep, task_id, dep_id}}`. Forces the
  caller to either include the dep task or remove the reference —
  silently treating unknown deps as satisfied would mask real bugs.
  """

  @type task_id :: String.t() | atom()
  @type task_spec :: {task_id(), [task_id()]}
  @type wave :: [task_id()]

  @doc """
  Plan execution waves for a list of `{task_id, depends_on}` pairs.

  Returns `{:ok, [wave1, wave2, ...]}` on success or
  `{:error, term}` on detected cycles / unknown deps.

  Task ordering within each wave is deterministic: tasks are
  emitted in the same order they appeared in the input list. This
  is important for reproducibility — a downstream dispatcher might
  prefer to schedule expensive tasks first, but that's a policy
  decision above this layer.
  """
  @spec topological_waves([task_spec()]) ::
          {:ok, [wave()]} | {:error, {:cycle, [task_id()]} | {:unknown_dep, task_id(), task_id()}}
  def topological_waves(tasks) when is_list(tasks) do
    with :ok <- check_unknown_deps(tasks) do
      ids = Enum.map(tasks, &elem(&1, 0))
      deps_map = Map.new(tasks)
      build_waves(ids, deps_map, MapSet.new(), [])
    end
  end

  @doc """
  Given a task list and a set of already-completed task ids, return
  the list of task ids that are ready to dispatch NOW: their
  dependencies are all in `completed`, and they themselves are not
  in `completed`. Order matches input order.

  Use to drive incremental dispatch: after each wave's tasks
  complete, merge them into the completed set and call again for
  the next batch.
  """
  @spec next_ready([task_spec()], MapSet.t(task_id())) :: [task_id()]
  def next_ready(tasks, completed) when is_list(tasks) do
    Enum.flat_map(tasks, fn {id, deps} ->
      cond do
        MapSet.member?(completed, id) -> []
        Enum.all?(deps, &MapSet.member?(completed, &1)) -> [id]
        true -> []
      end
    end)
  end

  # ── Internals ─────────────────────────────────────────────────────

  defp check_unknown_deps(tasks) do
    known = MapSet.new(tasks, &elem(&1, 0))

    Enum.reduce_while(tasks, :ok, fn {id, deps}, _acc ->
      case Enum.find(deps, fn d -> not MapSet.member?(known, d) end) do
        nil -> {:cont, :ok}
        bad -> {:halt, {:error, {:unknown_dep, id, bad}}}
      end
    end)
  end

  # Build waves greedily: each iteration picks every task whose deps
  # are all in `completed`. If a pass yields zero new tasks while
  # uncompleted tasks remain, we have a cycle.
  defp build_waves(remaining, deps_map, completed, acc) do
    {ready, still} = partition_ready(remaining, deps_map, completed)

    cond do
      ready == [] and still == [] ->
        {:ok, Enum.reverse(acc)}

      ready == [] and still != [] ->
        # No progress AND uncompleted tasks → cycle among `still`
        {:error, {:cycle, still}}

      true ->
        new_completed = Enum.reduce(ready, completed, &MapSet.put(&2, &1))
        build_waves(still, deps_map, new_completed, [ready | acc])
    end
  end

  defp partition_ready(ids, deps_map, completed) do
    Enum.split_with(ids, fn id ->
      deps = Map.fetch!(deps_map, id)
      Enum.all?(deps, &MapSet.member?(completed, &1))
    end)
  end
end

File: crates/worg-agent/src/scheduler.rs
//! Topological wave-planner for task DAGs.
//!
//! Pure data transformation — no concurrency primitives, no Tokio
//! task-spawning here. The Scheduler's role is the *plan*; the
//! integration that *dispatches* a wave via `tokio::spawn` /
//! `JoinSet` lives at the Loop edge.
//!
//! Mirrors Elixir's `WorgAgent.Scheduler` field-for-field at the
//! contract level so a single test corpus can validate both
//! runtimes (Phase 7).
//!
//! ## Example
//!
//! ```
//! use worg_agent::scheduler::{topological_waves, TaskSpec};
//!
//! let tasks: Vec<TaskSpec> = vec![
//!     ("research".into(), vec![]),
//!     ("script".into(), vec!["research".into()]),
//!     ("velocity".into(), vec!["research".into()]),
//!     ("storyboard".into(), vec!["script".into(), "velocity".into()]),
//!     ("shots".into(), vec!["storyboard".into()]),
//! ];
//! let waves = topological_waves(&tasks).unwrap();
//! assert_eq!(waves.len(), 4);
//! assert_eq!(waves[0], vec!["research"]);
//! assert_eq!(waves[1], vec!["script", "velocity"]);
//! ```

use std::collections::{BTreeMap, BTreeSet};

use thiserror::Error;

/// `(task_id, depends_on)`. Order in the input list is preserved
/// inside each emitted wave (deterministic output for reproducibility).
pub type TaskSpec = (String, Vec<String>);

#[derive(Debug, Error, PartialEq, Eq)]
pub enum ScheduleError {
    /// At least one cycle among the remaining tasks. The returned
    /// vec is the set of task ids that could not be ordered.
    #[error("cycle detected among tasks: {0:?}")]
    Cycle(Vec<String>),
    /// A task lists a `depends_on` that isn't in the input task
    /// list. Silently treating unknown deps as satisfied would mask
    /// real wiring bugs.
    #[error("task {task:?} depends on unknown id {dep:?}")]
    UnknownDep { task: String, dep: String },
}

/// Plan execution waves for a list of `(task_id, depends_on)` pairs.
///
/// Returns `Ok(vec_of_waves)` where each wave is a `Vec<String>` of
/// task ids that can run concurrently. Wave N's tasks all have
/// their dependencies satisfied by waves 1..N-1.
pub fn topological_waves(tasks: &[TaskSpec]) -> Result<Vec<Vec<String>>, ScheduleError> {
    check_unknown_deps(tasks)?;

    let ids: Vec<String> = tasks.iter().map(|(id, _)| id.clone()).collect();
    let deps_map: BTreeMap<String, Vec<String>> = tasks.iter().cloned().collect();
    build_waves(ids, &deps_map)
}

/// Given a task list + a set of already-completed task ids, return
/// the list of ids ready to dispatch NOW: their deps are all in
/// `completed` and they themselves are not. Order matches the input
/// list so a dispatcher can reproduce wave decisions identically.
pub fn next_ready(tasks: &[TaskSpec], completed: &BTreeSet<String>) -> Vec<String> {
    tasks
        .iter()
        .filter_map(|(id, deps)| {
            if completed.contains(id) {
                return None;
            }
            if deps.iter().all(|d| completed.contains(d)) {
                Some(id.clone())
            } else {
                None
            }
        })
        .collect()
}

fn check_unknown_deps(tasks: &[TaskSpec]) -> Result<(), ScheduleError> {
    let known: BTreeSet<&str> = tasks.iter().map(|(id, _)| id.as_str()).collect();
    for (id, deps) in tasks {
        if let Some(bad) = deps.iter().find(|d| !known.contains(d.as_str())) {
            return Err(ScheduleError::UnknownDep {
                task: id.clone(),
                dep: bad.clone(),
            });
        }
    }
    Ok(())
}

fn build_waves(
    mut remaining: Vec<String>,
    deps_map: &BTreeMap<String, Vec<String>>,
) -> Result<Vec<Vec<String>>, ScheduleError> {
    let mut completed: BTreeSet<String> = BTreeSet::new();
    let mut acc: Vec<Vec<String>> = Vec::new();

    loop {
        let (ready, still): (Vec<String>, Vec<String>) =
            remaining.iter().cloned().partition(|id| {
                deps_map
                    .get(id)
                    .map(|deps| deps.iter().all(|d| completed.contains(d)))
                    .unwrap_or(true)
            });

        if ready.is_empty() && still.is_empty() {
            return Ok(acc);
        }
        if ready.is_empty() {
            return Err(ScheduleError::Cycle(still));
        }
        for r in &ready {
            completed.insert(r.clone());
        }
        acc.push(ready);
        remaining = still;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn t(id: &str, deps: &[&str]) -> TaskSpec {
        (id.to_string(), deps.iter().map(|s| s.to_string()).collect())
    }

    #[test]
    fn linear_chain_yields_one_task_per_wave() {
        let tasks = vec![t("a", &[]), t("b", &["a"]), t("c", &["b"])];
        let waves = topological_waves(&tasks).unwrap();
        assert_eq!(waves, vec![vec!["a"], vec!["b"], vec!["c"]]);
    }

    #[test]
    fn parallel_branches_collapse_into_one_wave() {
        let tasks = vec![
            t("research", &[]),
            t("script", &["research"]),
            t("velocity", &["research"]),
            t("storyboard", &["script", "velocity"]),
            t("shots", &["storyboard"]),
        ];
        let waves = topological_waves(&tasks).unwrap();
        assert_eq!(waves.len(), 4);
        assert_eq!(waves[1], vec!["script", "velocity"]);
    }

    #[test]
    fn cycle_is_detected_and_surfaced() {
        let tasks = vec![t("a", &["b"]), t("b", &["a"])];
        match topological_waves(&tasks).unwrap_err() {
            ScheduleError::Cycle(ids) => {
                assert!(ids.contains(&"a".to_string()));
                assert!(ids.contains(&"b".to_string()));
            }
            other => panic!("expected Cycle, got {other:?}"),
        }
    }

    #[test]
    fn unknown_dep_is_rejected() {
        let tasks = vec![t("a", &["ghost"])];
        match topological_waves(&tasks).unwrap_err() {
            ScheduleError::UnknownDep { task, dep } => {
                assert_eq!(task, "a");
                assert_eq!(dep, "ghost");
            }
            other => panic!("expected UnknownDep, got {other:?}"),
        }
    }

    #[test]
    fn input_order_is_preserved_inside_each_wave() {
        let tasks = vec![
            t("a", &[]),
            t("c", &[]),
            t("b", &[]), // wave 1 should be a, c, b — input order
            t("d", &["a", "c", "b"]),
        ];
        let waves = topological_waves(&tasks).unwrap();
        assert_eq!(waves[0], vec!["a", "c", "b"]);
        assert_eq!(waves[1], vec!["d"]);
    }

    #[test]
    fn next_ready_skips_completed_and_unmet() {
        let tasks = vec![t("a", &[]), t("b", &["a"]), t("c", &["a"])];
        let mut done = BTreeSet::new();
        assert_eq!(next_ready(&tasks, &done), vec!["a"]);
        done.insert("a".to_string());
        assert_eq!(next_ready(&tasks, &done), vec!["b", "c"]);
        done.insert("b".to_string());
        done.insert("c".to_string());
        assert!(next_ready(&tasks, &done).is_empty());
    }

    #[test]
    fn empty_input_returns_empty_plan() {
        let tasks: Vec<TaskSpec> = vec![];
        assert!(topological_waves(&tasks).unwrap().is_empty());
    }
}
